src/analysis/processing/qgsalgorithmrandomextract.cpp

```diff
@@ -104,37 +109,106 @@ QVariantMap QgsRandomExtractAlgorithm::processAlgorithm( const QVariantMap &para
 
   // initialize random engine
   std::random_device randomDevice;
-  const std::mt19937 mersenneTwister( randomDevice() );
-  const std::uniform_int_distribution<int> fidsDistribution( 0, count );
-
-  QVector< QgsFeatureId > fids( number );
-  std::generate( fids.begin(), fids.end(), bind( fidsDistribution, mersenneTwister ) );
+  std::mt19937 mersenneTwister( randomDevice() );
 
-  QHash< QgsFeatureId, int > idsCount;
-  for ( const QgsFeatureId id : fids )
+  if ( !noDuplicates )
```

Comment:
Hmm, this branch is quite broken too -- it assumes that feature ids begin at 0 or 1 and have no gaps. That's not a valid assumption! It'd be safer to ALWAYS use the "build list of all feature ids" logic regardless of whether or not we are selecting duplicates...

Reply:
Ok